### PR TITLE
PHPLIB-1514 Make data providers static

### PR DESCRIPTION
**Pull request by GromNaN**
Fix PHPLIB-1514

In order to upgrade to PHPUnit 10+, we need to make all data provider static.

### CODE REVIEW

File: tests/TestCase.php [hunk 1]
```diff
@@ -206,57 +206,57 @@ protected function getCollectionName(): string
...
+    protected static function getInvalidObjectValues(bool $includeNull = false): array
     {
         return [123, 3.14, 'foo', true, [], new stdClass(), ...($includeNull ? [null] : [])];
     }
 
-    protected function getInvalidDocumentCodecValues(): array
+    protected static function getInvalidDocumentCodecValues(): array
     {
-        return [123, 3.14, 'foo', true, [], new stdClass(), $this->createMock(Codec::class)];
+        return [123, 3.14, 'foo', true, [], new stdClass(), self::createStub(Codec::class)];
```
Comment (GromNaN) [inline tests/TestCase.php]: Stubs can be created statically, mock cannot.

Comment (GromNaN) [reply tests/TestCase.php]: We have to wait for PHPUnit 10 to use this function.

**Comment by jmikola** [comment]
Do you want to address the [six errors in this CI build](https://github.com/mongodb/mongo-php-library/actions/runs/10855550373/job/30128393975?pr=1404#step:5:209) first?

File: tests/GridFS/FunctionalTestCase.php [hunk 2]
```diff
@@ -53,7 +53,7 @@ protected function assertStreamContents(string $expectedContents, $stream): void
      *
      * @return resource
      */
-    protected function createStream(string $data = '')
+    protected static function createStream(string $data = '')
```
Comment (jmikola) [inline tests/GridFS/FunctionalTestCase.php:56]: Technically not a data provider but there's also no reason this method can't be static :+1:

File: tests/GridFS/WritableStreamFunctionalTest.php [hunk 3]
```diff
@@ -86,7 +86,7 @@ public function testWriteBytesCalculatesMD5($input, $expectedMD5): void
         $this->assertSameDocument(['md5' => $expectedMD5], $fileDocument);
     }
 
-    public function provideInputDataAndExpectedMD5()
+    public static function provideInputDataAndExpectedMD5()
```
Comment (jmikola) [inline tests/GridFS/WritableStreamFunctionalTest.php:89]: Expect a conflict with https://github.com/mongodb/mongo-php-library/pull/1398 later.

File: tests/TestCase.php [hunk 4]
```diff
@@ -206,57 +208,84 @@ protected function getCollectionName(): string
...
+                return true;
+            }
+
+            public function encode(mixed $value): mixed
+            {
+                return $value;
+            }
+        };
+        // @fixme: createStub can be called statically in PHPUnit 10
+        // $codec = self::createStub(Codec::class);
```
Comment (jmikola) [inline tests/TestCase.php:264]: I assume once you upgrade the PHPUnit 10, the above anonymous class can be replaced with the single `createStub()` call?

Should this be tracked in a sub-task within [PHPLIB-1369](https://jira.mongodb.org/browse/PHPLIB-1369), or do you plan on addressing it in the main ticket?

**Review by jmikola — APPROVED** [review]
Some questions but changes LGTM.

File: tests/GridFS/FunctionalTestCase.php (hunk 2 above)
Comment (GromNaN) [reply tests/GridFS/FunctionalTestCase.php:56]: I made it static because It's used in a data provider.

File: tests/TestCase.php (hunk 4 above)
Comment (GromNaN) [reply tests/TestCase.php:264]: Tracked: https://jira.mongodb.org/browse/PHPLIB-1516

**Pull request merged**